man/mhparam.man
.TH MHPARAM %manext1% 2017-02-18 "%nmhversion%"
.
.\" %nmhwarning%
.
.SH NAME
mhparam \- print nmh profile and context components
.SH SYNOPSIS
.HP 5
.na
.B mhparam
.RB [ \-help ]
.RB [ \-version ]
.RI [ components ]
.RB [ \-all ]
.RB [ \-component " | " \-nocomponent ]
.RB [ \-debug ]
.ad
.SH DESCRIPTION
.B mhparam
writes the value of the specified profile component to the standard
output, separated by newlines.  Component name matching is
case-insensitive.  If the profile component is not present, the
default value (or nothing, if there is no default) is printed.
.PP
If the switch
.B \-component
is given, then the component name is displayed along with the profile
components value.  This can be disabled with the switch
.BR \-nocomponent .
.PP
If more than one component is specified in the
.I components
list, then the switch
.B \-component
is on by default.  If only one component is specified, then the switch
.B \-nocomponent
is on by default.
.PP
If
.B \-all
is specified, then all components in the nmh profile are displayed
and other arguments are ignored.
.PP
.B mhparam
can provide other information, such as the
.B nmh
version identifier, the locations of the nmh
.IR bindir ,
.IR libexecdir ,
.IR etcdir ,
and
.I docdir
install directories, all
.I proc
settings, and the
.I datalocking
method described in
.IR mh\-profile (5).
Information about the configuration of the
.B nmh
installation is also available, including the default mail spool
locking method.  To test whether the configuration includes
support for SASL, TLS,
.IR iconv (3),
and oauth, the following commands, respectively, will output a non-null
string to indicate support:
.PP
.RS 5
.nf
mhparam\0sasl
mhparam\0tls
mhparam\0iconv
mhparam\0oauth
.fi
.RE
.PP
Conversely, no output means that the specified support is not included.
The
.B \-debug
switch displays all such other information available from
.BR mhparam .
.PP
Examples:
.PP
.RS 5
.nf
.ta \w'AliasFile:'u+2n
$ mhparam path
Mail

$ mhparam mhlproc
%nmhlibexecdir%/mhl

$ mhparam \-component Path
Path: Mail

$ mhparam AliasFile rmmproc
AliasFile: aliases
rmmproc: rmmproc

$ mhparam \-nocomponent AliasFile rmmproc
aliases
rmmproc

$ mhparam path nonexistent context
Path: Mail
context: context

$ mhparam version
%nmhversion%

$ mhparam etcdir
%nmhetcdir%

$ mhparam libexecdir
%nmhlibexecdir%

$ mhparam spoollocking
%default_locking%

$ mhparam datalocking
fcntl

$ mhparam sasl
cyrus_sasl

$ mhparam tls
tls

$ mhparam iconv
iconv

$ mhparam oauth
oauth
.fi
.RE
.PP
.B mhparam
is also useful in backquoted operations:
.PP
.RS 5
.nf
$ fgrep cornell.edu \`mhpath +\`/\`mhparam aliasfile\`
.fi
.SH "EXIT STATUS"
The exit status is zero if all components are found,
or one if any are missing.
.
.SH FILES
.fc ^ ~
.nf
.ta \w'ExtraBigProfileName  'u
^$HOME/.mh_profile~^The user profile
^or $MH~^Rather than the standard profile
^<mh-dir>/context~^The user context
^or $MHCONTEXT~^Rather than the standard context
.fi
.SH "PROFILE COMPONENTS"
.fc ^ ~
.nf
.ta 2.4i
.ta \w'ExtraBigProfileName  'u
^context:~^To determine the context file
.fi
.SH "SEE ALSO"
.IR mh-profile (5)
.SH DEFAULTS
.nf
.RB ` \-nocomponent "' if only one component is specified"
.RB ` \-component "' if more than one component is specified"
.RB ` components "' defaults to none"
.fi
.SH CONTEXT
None


NEWS
# NEWS - Release notes for nmh 1.7
#

Welcome to nmh, the new version of the classic MH mail handling system.
It's been over three years since the last release of nmh, and there have
been a number of significant changes since the last release.  Long-time
MH and nmh uses should read careful the NOTEABLE CHANGES section, as there
are some significant changes to nmh behavior.  Otherwise, please see the
README and INSTALL files for help on getting started with nmh.

---------------
NOTABLE CHANGES
---------------

The largest notable changes in the 1.7 release are:

- Complete unification of network security support.  All network protocols
  (currently, POP and SMTP) have been refactored to use a common set of
  security routines.  This means all protocols support all SASL mechanisms
  (via the Cyrus-SASL library) and TLS.  TLS support has been strengthened
  to perform certificate name validation and to require TLS 1.1 as a
  minimum protocol.  Also, all protocols can make use of the OAuth2/XOAUTH
  SASL mechanism, which is supported by gmail.
- send(1) now supports adding switches to post(8) based on the address or
  domain of the email address in the From: header; this more easily allows
  users to support multiple identities.
- A generic facility for passing arguments to filter programs in repl(1)
  by use of the -convertargs switch.
- Native support for the manipulation of iCalendar requests; see mhical(1)
  for more details.

------------
NEW FEATURES
------------

The following are new features for the 1.7 release of nmh:

- When building from source, configure will derive ${prefix} from an existing
  nmh installation if it finds one in your $PATH.
- Added welcome message when nmh detects that its version changed.
- The default locations for configuration files and support binaries
  have been changed.  Configuration files now install into ${sysconfdir}/nmh,
  and support binaries are placed in ${libexecdir}/nmh.  If you are upgrading
  an existing installation you should look for old configuration files in
  ${sysconfdir} and merge any local customizations into the new files in
  ${sysconfdir}/nmh, then remove the old files.  ${libdir} will also contain
  obsolete support programs that should be removed.
- All TLS connections now perform certificate validation (including hostname
  matching) by default; can be disabled on a per-application basis.
- post now defaults to port 587 on 'smtp' message submission.
- A value of 0 for the width switch of scan(1), inc(1), ap(1), dp(1),
  fmttest(1), and mhl(1) now means as many characters as the format
  engine can produce [Bug #15274].  That amount is limited by internal
  buffers.
- If a component has trailing whitespace, e.g., body:component="> ",
  mhl now trims that whitespace off when filtering blank text lines.
- An "rtrim" flag has been added to mhl to remove any trailing
  whitespace from filtered text lines.  A corresponding "nortrim" flag
  has also been added.
- Added getmymbox and getmyaddr mh-format(5) function escapes.
- New -[no]changecur, -fixtype, -decodetypes, and -[no]crlflinebreaks switches
  have been added to mhfixmsg(1).
- mhfixmsg now removes an extraneous trailing semicolon from header
  parameter lists.
- Added -convertargs switch to repl(1), to pass arguments to programs
  specified in the user's profile or mhn.defaults to convert message
  content.
- Added mhical(1), to display, reply to, and cancel iCalendar (RFC 5545)
  event requests.
- Added multiply mh-format(5) function.
- "mhparam bindir" prints the path to the directory containing the public
  executables (${bindir}).
- New "-prefer" switch for mhshow (and mhlist and mhshow), to allow specifying
  the preferred content types to show, if present in a multipart alternative.
- mh-format now has %(kilo) and %(kibi) functions, to allow printing
  numbers with SI or IEC quantities, e.g. "10K", "2.3Mi".
- Support for the -sendmail flag to send/post to change the sendmail
  binary when using the sendmail/pipe MTS.
- Added support to send(1) to specify switches to post(1) based on address or
  domain name in From: header line in message draft.
- post(8) -snoop now attempts to decode base64-encoded SMTP traffic.
- folder(1) -nocreate now prints a warning message for a non-existent folder.
- mhfixmsg(1) now allows -decodetext binary, though 8bit is still the default.
- inc(1) and msgchk(1) now support TLS encryption natively.
- All network protocols support the XOAUTH authentication mechanism.
- Support for SMTPUTF8 (RFC 6531) has been added.  mhshow(1) already supported
  RFC 6532, assuming all 8-bit message header field bodies are UTF-8 and use
  of a UTF-8 locale.
- mhfixmsg now replaces RFC 2047 encoding with RFC 2231 encoding of name and
  filename parameters in Content-Type and Content-Disposition headers,
  respectively.
- If a message body contains 8-bit bytes, post(8) uses SMTP 8BITMIME if the
  server supports it.  If not, post fails with a message to the user to
  encode the message for 7-bit transport.
- Fewer lseek(2)s will be used when reading headers in the common case.
- ./configure's --enable-debug has been removed;  it did nothing.
- configure now defaults to enabling each of TLS and Cyrus SASL if the
  necessary headers and libraries are found.
- Moved build_nmh to top-level directory.
- Better error reporting for connections to network services.

---------
BUG FIXES
---------

- The format scanner no longer subtracts 1 from the width.  This has the
  effect of no longer counting the trailing newline in the output of
  scan(1), inc(1), and the other programs that rely on the format scanner.
- The first character of some very short (less than 4 characters) message
  bodies is no longer dropped.
- Single-character headers can be reliably formatted, etc., instead of
  apparently being missing.
- mhfixmsg now adds a Content-Transfer-Encoding header at the message level,
  if needed after decoding text parts.

uip/mhparam.c
/* mhparam.c -- print mh_profile values
 *
 * Originally contributed by
 * Jeffrey C Honig <Jeffrey_C_Honig@cornell.edu>
 *
 * This code is Copyright (c) 2002, by the authors of nmh.  See the
 * COPYRIGHT file in the root directory of the nmh distribution for
 * complete copyright information.
 */

#include <h/mh.h>
#include <h/mts.h>
#include "h/done.h"
#include <h/utils.h>

#define MHPARAM_SWITCHES \
    X("components", 0, COMPSW) \
    X("nocomponents", 0, NCOMPSW) \
    X("all", 0, ALLSW) \
    X("version", 0, VERSIONSW) \
    X("help", 0, HELPSW) \
    X("debug", 5, DEBUGSW) \

#define X(sw, minchars, id) id,
DEFINE_SWITCH_ENUM(MHPARAM);
#undef X

#define X(sw, minchars, id) { sw, minchars, id },
DEFINE_SWITCH_ARRAY(MHPARAM, switches);
#undef X

extern char *mhbindir;
extern char *mhlibexecdir;
extern char *mhetcdir;
extern char *mhdocdir;

static char *sbackup = BACKUP_PREFIX;

static char *datalocking = "fcntl";
static char *localmbox = "";
static int localmbox_primed = 0;

extern char *spoollocking;

static char *sasl =
#ifdef CYRUS_SASL
    "cyrus_sasl";
#else
    "";
#endif

static char *tls =
#ifdef TLS_SUPPORT
    "tls";
#else
    "";
#endif

static char *mimetypeproc =
#ifdef MIMETYPEPROC
    MIMETYPEPROC;
#else
    "";
#endif

static char *mimeencodingproc =
#ifdef MIMEENCODINGPROC
    MIMEENCODINGPROC;
#else
    "";
#endif

static char *iconv =
#ifdef HAVE_ICONV
    "iconv";
#else
    "";
#endif

static char *oauth =
#ifdef OAUTH_SUPPORT
    "oauth";
#else
    "";
#endif

struct proc {
    char *p_name;
    char **p_field;
};

static struct proc procs [] = {
     { "context",          &context },
     { "mh-sequences",     &mh_seq },
     { "buildmimeproc",    &buildmimeproc },
     { "fileproc",         &fileproc },
     { "foldprot",         &foldprot },
     { "formatproc",	   &formatproc },
     { "incproc",          &incproc },
     { "lproc",            &lproc },
     { "mailproc",         &mailproc },
     { "mhlproc",          &mhlproc },
     { "mimetypeproc",     &mimetypeproc },
     { "mimeencodingproc", &mimeencodingproc },
     { "moreproc",         &moreproc },
     { "msgprot",          &msgprot },
     { "packproc",         &packproc },
     { "postproc",         &postproc },
     { "rmmproc",          &rmmproc },
     { "sendproc",         &sendproc },
     { "showmimeproc",     &showmimeproc },
     { "showproc",         &showproc },
     { "version",          &version_num },
     { "whatnowproc",      &whatnowproc },
     { "whomproc",         &whomproc },
     { "bindir",           &mhbindir },
     { "libexecdir",       &mhlibexecdir },
     { "etcdir",           &mhetcdir },
     { "docdir",           &mhdocdir },
     { "localmbox",	   &localmbox },
     { "sbackup",          &sbackup },
     { "datalocking",      &datalocking },
     { "spoollocking",     &spoollocking },
     { "iconv",		   &iconv },
     { "oauth",		   &oauth },
     { "sasl",             &sasl },
     { "tls",              &tls },
     { NULL,               NULL },
};


/*
 * static prototypes
 */
static char *p_find(char *) PURE;


int
main(int argc, char **argv)
{
    int i, compp = 0;
    bool missed;
    int all = 0, debug = 0;
    int components = -1;
    char *cp, buf[BUFSIZ], **argp;
    char **arguments, *comps[MAXARGS];

    if (nmh_init(argv[0], 2)) { return 1; }

    arguments = getarguments (invo_name, argc, argv, 1);
    argp = arguments;

    while ((cp = *argp++)) {
	if (*cp == '-') {
	    switch (smatch (++cp, switches)) {
		case AMBIGSW: 
		    ambigsw (cp, switches);
		    done (1);
		case UNKWNSW: 
		    adios (NULL, "-%s unknown", cp);

		case HELPSW: 
		    snprintf (buf, sizeof(buf), "%s [profile-components] [switches]",
			invo_name);
		    print_help (buf, switches, 1);
		    done (0);
		case VERSIONSW:
		    print_version(invo_name);
		    done (0);

		case COMPSW:
		    components = 1;
		    break;
		case NCOMPSW:
		    components = 0;
		    break;

		case ALLSW:
		    all = 1;
		    break;

		case DEBUGSW:
		    debug = 1;
		    break;
	    }
	} else {
	    comps[compp++] = cp;
	    if (strcmp("localmbox", cp) == 0 && ! localmbox_primed) {
		localmbox = getlocalmbox();
		localmbox_primed = 1;
	    }
	}
    }

    if (all) {
        struct node *np;

	if (compp)
	    inform("profile-components ignored with -all");

	if (components >= 0)
	    inform("-%scomponents ignored with -all",
		   components ? "" : "no");

	/* Print all entries in context/profile list.  That does not
	   include entries in mts.conf, such as spoollocking. */
	for (np = m_defs; np; np = np->n_next)
	    printf("%s: %s\n", np->n_name, np->n_field);

    }

    if (debug) {
	struct proc *ps;

	/* In case datalocking was set in profile. */
	if ((cp = context_find("datalocking"))) { datalocking = cp; }

	/* In case spoollocking was set in mts.conf. */
	mts_init();

	/* Also set localmbox here */
	if (! localmbox_primed) {
	    localmbox = getlocalmbox();
	    localmbox_primed = 1;
	}

	/*
	 * Print the current value of everything in
	 * procs array.  This will show their current
	 * value (as determined after context is read).
         */
	for (ps = procs; ps->p_name; ps++)
	    printf ("%s: %s\n", ps->p_name, FENDNULL(*ps->p_field));

    }

    missed = false;
    if (! all) {
        if (components < 0)
	    components = compp > 1;

	for (i = 0; i < compp; i++)  {
	    char *value;

	    if (! strcmp ("spoollocking", comps[i])) {
		/* In case spoollocking was set in mts.conf. */
		mts_init();
            }

	    value = context_find (comps[i]);
	    if (!value)
		value = p_find (comps[i]);
	    if (value) {
	        if (components)
		    printf("%s: ", comps[i]);

		puts(value);
	    } else
                missed = true;
	}
    }

    done(missed);
}


static char *
p_find(char *str)
{
    struct proc *ps;

    for (ps = procs; ps->p_name; ps++)
	if (!strcasecmp (ps->p_name, str))
	    return *ps->p_field;

    return NULL;
}
